File: Leetcode/cp36_valid_sudoku/Solution.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] != '.') {
                    char num = board[i][j];
                    if (isValidRowColumn(board, num, i, j) ) {
                        continue;
                    } else {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    
    private boolean isValidRowColumn(char[][] board, char num, int row, int col) {
        for (int i = 0; i < 9; i++) {
            // The number in the row or column should not be the current position
            if ((board[row][i] == num && i != col) || (board[i][col] == num && i != row)) {
                return false;
            }
            int r = 3 * (row / 3) + i / 3;
            int c = 3 * (col / 3) + i % 3;
            // The number in the 3X3 grid should not be the current position
            if (board[r][c] == num && (r != row && c != col)) {
                return false;
            }
        }
        return true;
    }
}

File: Leetcode/cp36_valid_sudoku/Solution.py
class Solution:
    # Leetcode Problem: https://leetcode.com/problems/valid-sudoku/description/
    #
    # Time complexity: O(1) since the input length is always fixed
    # Space complexity: O(1)
    def is_each_row_valid(self, board: [list[list[str]]]) -> bool:
        for row in board:
            if not self.is_valid_unit(row):
                return False
        return True
    
    def is_each_col_valid(self, board: [list[list[str]]]) -> bool:
        for col in zip(*board):
            if not self.is_valid_unit(col):
                return False
        return True
    
    def is_each_square_valid(self, board: [list[list[str]]]) -> bool:
        for i in (0, 3, 6):
            for j in (0, 3, 6):
                square = [
                    board[x][y]
                    for x in range(i, i + 3) for y in range(j, j + 3)
                ]
                if not self.is_valid_unit(square):
                    return False
        return True
    
    def is_valid_unit(self, unit: list[str]) -> bool:
        nums = [num for num in unit if num != "."]
        return len(set(nums)) == len(nums)

    def isValidSudoku(self, board: list[list[str]]) -> bool:
        return (
            self.is_each_row_valid(board) and 
            self.is_each_col_valid(board) and 
            self.is_each_square_valid(board)
        )
